### ISSUE TITLE
Steps to test a new game and crash in Perception

### ISSUE THREAD

**Issue by samaust** [cybereality/Perception#66]
I'm testing Perception master on a Direct3D9 simple app I made using an old D3D9 tutorial and a sample in the Microsoft DirectX SDK (June 2010). It allows to quickly do basic testing without the long load time of a full game. I used that app when I worked on [Triditizer](https://github.com/samaust/Triditizer). I've uploaded this D3D9 test app on GitHub under the name [dxgraphics18](https://github.com/samaust/dxgraphics18). Perception crashes in dxgraphics18 and I'm trying to figure out why. I'm getting stack overflow. It looks like it gets stuck into a loop calling three functions :
1. HRESULT WINAPI DataGatherer::SetVertexShaderConstantF(UINT StartRegister,CONST float\* pConstantData,UINT Vector4fCount)
2. HRESULT WINAPI D3DProxyDevice::SetVertexShaderConstantF(UINT StartRegister,CONST float\* pConstantData,UINT Vector4fCount)
3. HRESULT WINAPI D3D9ProxyStateBlock::SelectAndCaptureStateVSConst(UINT StartRegister,CONST float\* pConstantData,UINT Vector4fCount)

The third function will call the first one and it goes like this until it crashes.

Could someone with experience in programming for Perception :
- advise me on what are the configurations steps for testing a new game (settings to use in config.xml, profiles.xml, shader_analyser.cfg, etc.)
- give a try to dxgraphics18 with Perception and see it they can find out why it crashes

**Comment by samaust** [comment]
I think I found a solution. There were two problems.

SelectAndCaptureStateVSConst calls

```
HRESULT result = m_pOwningDevice->SetVertexShaderConstantF(StartRegister, pConstantData, Vector4fCount);
```

pOwningDevice comes from 

```
m_pCapturingStateTo = new D3D9ProxyStateBlock(NULL, this, D3D9ProxyStateBlock::Cap_Type_Selected, m_currentRenderingSide == vireio::Left);
```

I think SelectAndCaptureStateVSConst  meant to call SetVertexShaderConstantF in BaseDirect3DDevice9 so I modified the line

```
HRESULT result = m_pOwningDevice->SetVertexShaderConstantF(StartRegister, pConstantData, Vector4fCount);
```

to

```
HRESULT result = (*(D3DProxyDevice*)m_pOwningDevice).BaseDirect3DDevice9::SetVertexShaderConstantF(StartRegister, pConstantData, Vector4fCount);
```

Someone more knowledgeable with the inner working of Perception will have to see if this is the proper fix.

After that, I got a crash in UnWrapTexture when trying to use a released texture. The game calls

```
IDirect3DTexture9 *pTex;
if( SUCCEEDED( D3DXCreateTextureFromFile( pd3dDevice, strTexture, &pTex ) ) )
{
    // Obtain the base texture interface
    pTex->QueryInterface( IID_IDirect3DBaseTexture9, (LPVOID*)&m_pTextures[i] );
    // Release the specialized instance
    pTex->Release();
}
```

QueryInterface ends up calling D3D9ProxyTexture::QueryInterface which is defined this way

```
HRESULT WINAPI D3D9ProxyTexture::QueryInterface(REFIID riid, LPVOID* ppv)
{
	SHOW_CALL("D3D9ProxyTexture::QueryInterface");
	/* IID_IDirect3DTexture9 */
	/* {85C31227-3DE5-4f00-9B3A-F11AC38C18B5} */
	IF_GUID(riid,0x85c31227,0x3de5,0x4f00,0x9b,0x3a,0xf1,0x1a)
	{	
		*ppv=(LPVOID)this;
		this->AddRef();
		return S_OK;
	}

	/* IID_IDirect3DSurface9 */
	/* {0CFBAF3A-9FF6-429a-99B3-A2796AF8B89B} */
	IF_GUID(riid,0x0cfbaf3a,0x9ff6,0x429a,0x99,0xb3,0xa2,0x79)
		return this->GetSurfaceLevel(0,(IDirect3DSurface9**)ppv);
	
	return m_pActualTexture->QueryInterface(riid, ppv);
}
```

Notice that the reference is incremented if the GUID is equal to IID_IDirect3DTexture9. The problem is that the game passed a IID_IDirect3DBaseTexture9 GUID so D3D9ProxyTexture::QueryInterface does not increase the reference count and the Release call releases the texture. Adding the following code 

```
/* IID_IDirect3DBaseTexture9 */
	/* {580CA87E-1D3C-4d54-991D-B7D3E3C298CE} */
	IF_GUID(riid, 0x580ca87e, 0x1d3c, 0x4d54, 0x99, 0x1d, 0xb7, 0xd3, 0xe3, 0xc2, 0x98, 0xce)
	{
		*ppv = (LPVOID)this;
		this->AddRef();
		return S_OK;
	}
```

increments the reference, the textures are not released and the game completely loads and I'm seeing the menu.

I'm seeing a problem with the menus. When I move up or down in the menu, in one eye, there's a pink background that stays behind the menu. I'll do some more testing to try to figure out why it's happenning.

**Comment by samaust** [comment]
Here's a screenshot of the issue with the menu
![issue_menu](https://cloud.githubusercontent.com/assets/10716165/13382667/9e094f28-de48-11e5-81d3-3d264e552541.jpg)

**Comment by EisernSchild** [comment]
Hi Samuel !

Sorry for the late answer, i was quite busy integrating OSVR to the new
driver architecture.

1) Yes, you are fully right, the member m_pOwningDevice in
D3DProxyStateblock is actually a wrapped device (D3D9ProxyDevice) and not
an actual device (IDirect3DDevice9), as defined. You see that the original
code in D3DProxyDevice::CreateStateBlock() :

*ppSB = new D3D9ProxyStateBlock(pActualStateBlock, this, capType,
m_currentRenderingSide == vireio::Left);

So, yes, your fix should work. The other way would be to provide the actual
device instead of the current (this) when creating the class in the line
below.

For our excuse : The State Block class was written by Chris Drain back in
2013, short before i entered the project. We never really tested it since
we never faced a game that actually had troubles with it.

2) Also, your QueryInterface fix should work as well the fixes we already
added to that method (here in D3D9ProxyTexture::QueryInterface as well as
in D3DProxyDevice::QueryInterface).
I added these QueryInterface fixes back in 2013 since Skyrim was crashing
exactly due to the same reason that your game now crashed calling this
method.

3) The image looks to me as if the left render target does not get cleared
for some reason (whereas the right one gets cleared properly). So, whenever
you move the pink menu lines up/down these menu lines never get cleared, i
guess.


So, great work so far from you !! I write you another mail including the
whole team to tell you what we exactly currently do.

Regards, Denis


2016-02-29 0:25 GMT+01:00 Samuel Austin <notifications@github.com>:

**Comment by EisernSchild** [comment]
Samuel, could you contact us directly via info@mtbs3d.com ? We do not have
an email address from you and it would be great if we could meet up via
mail.

Neil Schneider is the host of MTBS, he will organize a meet-up mail
including the team.

2016-03-01 17:45 GMT+01:00 Denis Reischl <eisernschild@gmail.com>:

**Comment by samaust** [comment]
After applying my fix to load the possible shader constant names in m_wvpMatrixConstantNames, the shader rule was filled by the built-in analyzer. I deleted manually two shaders that were not necessary from the shader rule file and I increased fWorldScaleFactor and 3D now works.

Now, I just need to figure out why one eye if not clearing properly and it will be working for me in this D3D9 sample. After that, I'll be ready to try it in real D3D9 games.
